File: src/ru/hudspeed/pro/MapVisualStyle.java
package ru.gpsantiradar.app;

/** Pure color decisions shared by camera coverage and the current-location marker. */
public final class MapVisualStyle {
    public static final int LOCATION_PRIMARY_COLOR = 0xFF1976D2;
    public static final int LOCATION_HIGHLIGHT_COLOR = 0xFF64B5F6;
    private static final int NIGHT_LOCATION_PRIMARY_COLOR = 0xFFFFBE00;
    private static final int NIGHT_LOCATION_HIGHLIGHT_COLOR = 0xFFFFD848;

    private static final int NORMAL_FILL_ALPHA = 0x26;
    private static final int NORMAL_STROKE_ALPHA = 0x4D;
    private static final int ACTIVE_FILL_ALPHA = 0x4D;
    private static final int ACTIVE_STROKE_ALPHA = 0x73;

    private MapVisualStyle() {}

    public static int locationPrimaryColor(boolean nightMode) {
        return nightMode ? NIGHT_LOCATION_PRIMARY_COLOR : LOCATION_PRIMARY_COLOR;
    }

    public static int locationHighlightColor(boolean nightMode) {
        return nightMode ? NIGHT_LOCATION_HIGHLIGHT_COLOR : LOCATION_HIGHLIGHT_COLOR;
    }

    public static Coverage coverage(int baseColor, long cameraId, long activeCameraId) {
        boolean active = cameraId >= 0L && cameraId == activeCameraId;
        return new Coverage(
                argb(baseColor, active ? ACTIVE_FILL_ALPHA : NORMAL_FILL_ALPHA),
                argb(baseColor, active ? ACTIVE_STROKE_ALPHA : NORMAL_STROKE_ALPHA));
    }

    private static int argb(int baseColor, int alpha) {
        return (alpha << 24) | (baseColor & 0x00FFFFFF);
    }

    public static final class Coverage {
        public final int fillColor;
        public final int strokeColor;

        private Coverage(int fillColor, int strokeColor) {
            this.fillColor = fillColor;
            this.strokeColor = strokeColor;
        }
    }
}

File: src/main/kotlin/ru/gpsantiradar/app/MapVisualStyle.kt
package ru.gpsantiradar.app

/** Pure color decisions shared by camera coverage and the current-location marker. */
object MapVisualStyle {
    val LOCATION_PRIMARY_COLOR: Int = 0xFF1976D2.toInt()
    val LOCATION_HIGHLIGHT_COLOR: Int = 0xFF64B5F6.toInt()
    private val nightLocationPrimaryColor = 0xFFFFBE00.toInt()
    private val nightLocationHighlightColor = 0xFFFFD848.toInt()
    private const val NORMAL_FILL_ALPHA = 0x26
    private const val NORMAL_STROKE_ALPHA = 0x4D
    private const val ACTIVE_FILL_ALPHA = 0x4D
    private const val ACTIVE_STROKE_ALPHA = 0x73

    fun locationPrimaryColor(nightMode: Boolean) =
        if (nightMode) nightLocationPrimaryColor else LOCATION_PRIMARY_COLOR

    fun locationHighlightColor(nightMode: Boolean) =
        if (nightMode) nightLocationHighlightColor else LOCATION_HIGHLIGHT_COLOR

    fun coverage(baseColor: Int, cameraId: Long, activeCameraId: Long): Coverage {
        val active = cameraId >= 0L && cameraId == activeCameraId
        return Coverage(
            argb(baseColor, if (active) ACTIVE_FILL_ALPHA else NORMAL_FILL_ALPHA),
            argb(baseColor, if (active) ACTIVE_STROKE_ALPHA else NORMAL_STROKE_ALPHA)
        )
    }

    private fun argb(baseColor: Int, alpha: Int) = alpha shl 24 or (baseColor and 0x00FFFFFF)

    class Coverage internal constructor(val fillColor: Int, val strokeColor: Int)
}
